### PR TITLE
bridge: Check integer bounds in dbus method args

### LINKED ISSUE #930: Explicitly raise integer size errors in JSON DBus bridge
So when we cast JSON integers to DBus types, we should be checking that tehy are in bonds for the destination type. The code here doesn't do that:

https://github.com/cockpit-project/cockpit/blob/master/src/agent/cockpitdbusjson.c#L351

We should produce an error (ie: `G_IO_ERROR_INVALID_DATA`) when that's not the case.

### PR DESCRIPTION
**Pull request by larskarlitski**
Fixes #930

### CODE REVIEW

File: src/bridge/cockpitdbusjson.c [hunk 1]
```diff
@@ -122,6 +122,30 @@ check_type (JsonNode *node,
   return TRUE;
 }
 
+static gboolean
+check_int_type (JsonNode *node,
+                const char *type_name,
```
Comment (martinpitt) [inline src/bridge/cockpitdbusjson.c]: As this is only ever being used for checking `GVariant` ranges, how about passing the `type` field directly and using `g_variant_get_type_string()` on it, instead of manually creating the type strings?

File: src/bridge/cockpitdbusjson.c [hunk 2]
```diff
@@ -476,37 +500,38 @@ parse_json (JsonNode *node,
...
         {
-          if (check_type (node, JSON_NODE_VALUE, G_TYPE_INT64, error))
+          if (check_int_type (node, "int64", G_MININT64, G_MAXINT64, error))
             return g_variant_new_int64 (json_node_get_int (node));
         }
       else if (g_variant_type_equal (type, G_VARIANT_TYPE_UINT64))
         {
-          if (check_type (node, JSON_NODE_VALUE, G_TYPE_INT64, error))
+          /* can never be larger than signed int64 */
+          if (check_int_type (node, "uint64", 0, G_MAXINT64, error))
```
Comment (martinpitt) [inline src/bridge/cockpitdbusjson.c]: This maintains current behaviour, but the comment doesn't really explain anything. It should say *why* this can't be bigger than `G_MAXINT64`. glib, D-Bus etc. certainly don't have this limitation, and JavaScript is completely odd (everything is a double, so with large numbers you'll lose precision but they are not bounded) so where does it come from?

My guess is that it merely comes from the signature and types in `check_int_type()`, as that by itself cannot accept guint64. I'd rather recommend using a separate `check_guint64_type()` for that, this seems more robust than arbitrarily breaking an official D-Bus data type.

**Comment by stefwalter** [comment]
Good to get this in now before we test on alternate architectures.

File: src/base1/test-dbus-common.js [hunk 3]
```diff
@@ -108,6 +108,39 @@ function common_dbus_tests(channel_options, bus_name)
...
+        testNumber('n', 0x7fff + 1);
+        testNumber('q', -1);
+        testNumber('q', 0xffff + 1);
+        testNumber('i', -0x80000000 -1);
+        testNumber('i', 0x7fffffff + 1);
+        testNumber('u', -1);
+        testNumber('u', 0xffffffff + 1);
+        testNumber('t', -1);
+
+        QUnit.start();
```
Comment (martinpitt) [inline src/base1/test-dbus-common.js]: Would be good to test the boundary cases which are supposed to work as well (like e. g. a large guint64).

File: src/bridge/cockpitdbusjson.c (hunk 1 above)
Comment (larskarlitski) [inline src/bridge/cockpitdbusjson.c]: I didn't do that because D-Bus' type strings are hard to remember, especially the integers. But I grant you that when dealing with D-Bus, you need to know them anyway and introducing different ones is weird. Changed in the update.

File: src/bridge/cockpitdbusjson.c (hunk 2 above)
Comment (larskarlitski) [inline src/bridge/cockpitdbusjson.c]: This is a limitation of json-glib, which only deals in `doubles` and `int64s`. I've clarified the comment.

File: src/base1/test-dbus-common.js (hunk 3 above)
Comment (larskarlitski) [inline src/base1/test-dbus-common.js]: I've added a couple of positive test cases, with numbers that are bigger than the next smaller type. Large `guint64` values aren't supported, because json-glib only gives us `gint64`. This might be something we need to address in the future, but these are mostly used for microsecond time stamps, so we've got a bit of time ;)

File: src/bridge/cockpitdbusjson.c (hunk 1 above)
Comment (martinpitt) [inline src/bridge/cockpitdbusjson.c]: Oh right, I thought `get_type_string()` would give you something more readable, sorry:
```sh
$ python3 -c 'from gi.repository import GLib; print(GLib.Variant.new_int64(1).get_type_string())'
x
```
That's of course less than ideal. So I'm fine with keeping the manual strings, sorry for not checking this more closely before!

File: src/base1/test-dbus-common.js (hunk 3 above)
Comment (larskarlitski) [inline src/base1/test-dbus-common.js]: Had to change this a bit because positive test cases means that the properties were actually set, which broke assumptions in other tests. I've now added an additional method just for this test.

File: src/bridge/cockpitdbusjson.c (hunk 1 above)
Comment (larskarlitski) [inline src/bridge/cockpitdbusjson.c]: Heh, I've changed it now, but I wouldn't mind changing it back.
